### DISCUSSION TITLE
Generic routable viewmodel

### DISCUSSION THREAD

**Discussion by AlienJust** [reactiveui/ReactiveUI#3440] [Q&A]
Hi, I'm writing Avalonia desktop app and I have VM that looks like:
`public class SampleViewModel<T> : ReactiveObject, IRoutableViewModel, ISampleViewModel {...}`

interface ISampleViewModel **is not generic**:
```
public interface ISampleViewModel : IRoutableViewModel, INotifyPropertyChanged
{
    string Greeting { get; }
}
```

also I have view for `ISampleViewModel`:
`public partial class SampleView : ReactiveUserControl<ISampleViewModel>`

in `app.axaml.cs` I register View for VM class:
`Locator.CurrentMutable.Register<IViewFor<ISampleViewModel>>(() => new SampleView());`

Then I try to navigate to it from my MainViewModel:
```
public ICommand NavigateToSampleVm => ReactiveCommand.Create(() =>
{
    Router.Navigate.Execute(new SampleViewModel<byte>())
    .Subscribe(vm => { Console.WriteLine(">> Current vm is " + ((vm == null) ? "NULL" : vm.GetType().FullName)); });
    Console.WriteLine("Navigated to sample view model");
});
```

But I don't see my view inside `RoutedViewHost` in `MainWindow` (there is only `DefaultContent` there).

If I remove generic part from VM class `SampleViewModel` then navigation works as expected.

![output](https://user-images.githubusercontent.com/739189/205672804-509ab124-b56f-4022-9739-d0ad4df1feb6.gif)

Is it possible to use generic view models for navigation somehow?

**Comment by glennawatson** [comment]
With Avalonia it's worth also highlighting it to their team. They have gitter chat. 

They produce their own rxui plugin so having both teams eyes on the problem can be worthwhile.

**Comment by AlienJust** [comment]
Tried this scenario for WPF. Got the same behavior.
`Locator.CurrentMutable.Register<IViewFor<SampleViewModel<byte>>>(() => new SampleView());` works well.

**Comment by AlienJust** [comment]
Hm, it may be related https://github.com/reactiveui/Splat.DI.SourceGenerator/issues/137

**Reply by glennawatson** [reply]
Unlikely. Splat source generator just auto registers stuff for you. Unlikely to be splat itself. Best bet is to ask over at Avalonia. The RoutedViewHost is their control and they handle their own routing and reactiveui integration.

**Comment by csillikd** [comment]
Any solution found?

**Reply by glennawatson** [reply]
Likely your best avenue for an answer is the Avalonia team. They manage their own reactiveui plugin.

**Reply by glennawatson** [reply]
Specifically the RoutedViewHost etc are all Avalonia components.
